### PR TITLE
feat(ci): remove actions-rs/cargo and use r7kamura/rust-problem-matchers

### LINKED ISSUE #3398: Move away from `actions-rs/cargo`
## Description
The one feature we are currently getting from this action is check annotations as part of the pull request. We can replace this with something like: https://github.com/r7kamura/rust-problem-matchers. There might be better options out there.

There was a feature-request for `dtolnay/rust-toolchain` to add this but it was denied: https://github.com/dtolnay/rust-toolchain/issues/50

## Motivation

For the same reason as moving away from `actions-rs/toolchain`. See #3352.

## Are you planning to do it yourself in a pull request?

No.

### PR DESCRIPTION
**Pull request by PopBogdan97**
## Description

Resolves https://github.com/libp2p/rust-libp2p/issues/3398.

## Notes

Removed also the `.cargo` folder with just the `.confing.toml` file since it was used just to run the command in `actions-rs/cargo`

## Links to any relevant issues

Related issue #3398 

## Open Questions

1. I am not sure if the `cargo clippy` command should check also the dependencies, because on this pull request it seems that does not check the dependencies, instead on the pull request I did on my own fork it seems that it does the checks on the dependencies.
Which should be the correct behavior?
(If you know the reason for the different behavior on the two repos could you explain it to me?)

2. I think there are no docs or tests on the workflows, do I need to check something else?

I am just starting to contribute to open source code thus any comments are welcome :) 

## Change checklist

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: .cargo/config.toml [hunk 1]
```diff
@@ -1,3 +0,0 @@
-[alias]
-# Temporary solution to have clippy config in a single place until https://github.com/rust-lang/rust-clippy/blob/master/doc/roadmap-2021.md#lintstoml-configuration is shipped.
-custom-clippy = "clippy --workspace --all-features --all-targets -- -A clippy::type_complexity -A clippy::pedantic -W clippy::used_underscore_binding -D warnings"
```
Comment (thomaseizinger) [inline .cargo/config.toml]: Please don't remove this. We want to retain the alias to make it easy for users to run the command locally.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -190,10 +190,10 @@ jobs:
         with:
           save-if: ${{ github.ref == 'refs/heads/master' }}
 
+      - uses: r7kamura/rust-problem-matchers@v1
```
Comment (thomaseizinger) [inline .github/workflows/ci.yml]: Can you also set this one up for the following jobs please:

- rustfmt
- rustdoc links
- feature matrix
- cross
- test

Essentially, all jobs where we run `cargo`.

Ordering wise, I'd prefer to to be directly after the `rust-toolchain` action.

**Comment by thomaseizinger** [comment]
> 1. I am not sure if the `cargo clippy` command should check also the dependencies, because on this pull request it seems that does not check the dependencies, instead on the pull request I did on my own fork it seems that it does the checks on the dependencies.
>        Which should be the correct behavior?

I am surprised there are any differences? In general, `cargo` will silence any warnings or lints in dependencies because there is nothing you can do about it (it is not your code) :)

**Review by thomaseizinger — CHANGES_REQUESTED** [review]
Two requests, otherwise LGTM!

File: .cargo/config.toml (hunk 1 above)
Comment (PopBogdan97) [reply .cargo/config.toml]: Sure

File: .github/workflows/ci.yml (hunk 2 above)
Comment (PopBogdan97) [reply .github/workflows/ci.yml]: Sure, I will do it.

**Comment by PopBogdan97** [comment]
> I am surprised there are any differences? In general, `cargo` will silence any warnings or lints in dependencies because there is nothing you can do about it (it is not your code) :)

In principle you are right, but it still does the check, this is why I was curious. I will post here two screenshot to show the difference:
Screenshot from 2023-02-28 15-25-54
Screenshot from 2023-02-28 15-26-16

File: .cargo/config.toml (hunk 1 above)
Comment (PopBogdan97) [reply .cargo/config.toml]: Done

File: .github/workflows/ci.yml (hunk 2 above)
Comment (PopBogdan97) [reply .github/workflows/ci.yml]: I added it for all the jobs that run cargo except from the ones that redirect the output to $GITHUB_OUTPUT

**Comment by thomaseizinger** [comment]
> > I am surprised there are any differences? In general, `cargo` will silence any warnings or lints in dependencies because there is nothing you can do about it (it is not your code) :)
> 
> In principle you are right, but it still does the check, this is why I was curious. I will post here two screenshot to show the difference:

Right, I see what you mean. What is happening here is that cargo needs to type-check all dependencies before it can type-check our code as it needs to internally build up the graph of types. That is why you can see "Checking XYZ" for our dependencies. If you run it with extra verbosity (`-vv` I believe), you will even see the warnings emitted from these crates.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -39,6 +39,8 @@ jobs:
         with:
           toolchain: ${{ steps.parse-msrv.outputs.version }}
 
+      - uses: r7kamura/rust-problem-matchers@v1
```
Comment (thomaseizinger) [inline .github/workflows/ci.yml]:
```suggestion
      - uses: r7kamura/rust-problem-matchers@d58b70c4a13c4866d96436315da451d8106f8f08 @ v1
```

For most third-party actions, we pin the hash to lower the risk of a supply-chain attack. Please use the hash everywhere.

Comment (thomaseizinger) [reply .github/workflows/ci.yml]: Dependabot will send PRs for bumping these.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -186,14 +194,14 @@ jobs:
...
+
       - uses: Swatinem/rust-cache@359a70e43a0bb8a13953b04a90f76428b4959bb6 # v2.2.0
         with:
           save-if: ${{ github.ref == 'refs/heads/master' }}
 
       - name: Run cargo clippy
-        uses: actions-rs/cargo@844f36862e911db73fe0815f00a4a2602c279505 # v1.0.3
-        with:
-          command: custom-clippy # cargo alias to allow reuse of config locally
+        run: cargo clippy --workspace --all-features --all-targets -- -A clippy::type_complexity -A clippy::pedantic -W clippy::used_underscore_binding -D warnings
```
Comment (thomaseizinger) [inline .github/workflows/ci.yml]:
```suggestion
        run: cargo custom-clippy # cargo alias to allow reuse of config locally
```

Please use the alias to make sure we run the same config locally as in CI.

**Review by thomaseizinger — CHANGES_REQUESTED** [review]
Two more requests, otherwise LGTM! :)

**Comment by PopBogdan97** [comment]
Done. Please let me know if there is something else I missed and thank you for your help :)

**Review by thomaseizinger — DISMISSED** [review]
Excellent, thank you!

**Review by thomaseizinger — APPROVED** [review]

**Pull request merged**